src/gluonts/model/forecast_generator.py

```diff
@@ -0,0 +1,222 @@
+
+import logging
+from typing import Any, Callable, Iterator, List, Optional
+
+# Third-party imports
+import mxnet as mx
+import numpy as np
+
+# First-party imports
+from gluonts.distribution import Distribution, DistributionOutput
+from gluonts.core.component import validated
+from gluonts.dataset.common import DataEntry
+from gluonts.dataset.field_names import FieldName
+from gluonts.dataset.loader import InferenceDataLoader
+from gluonts.model.forecast import (
+    Forecast,
+    SampleForecast,
+    QuantileForecast,
+    DistributionForecast,
+)
+
+
+OutputTransform = Callable[[DataEntry, np.ndarray], np.ndarray]
+BlockType = mx.gluon.Block
+
+
+LOG_CACHE = set([])
+
+
+def log_once(msg):
+    global LOG_CACHE
+    if msg not in LOG_CACHE:
+        logging.info(msg)
+        LOG_CACHE.add(msg)
+
+
+def _extract_instances(x: Any) -> Any:
+    """
+    Helper function to extract individual instances from batched
+    mxnet results.
+
+    For a tensor `a`
+      _extract_instances(a) -> [a[0], a[1], ...]
+
+    For (nested) tuples of tensors `(a, (b, c))`
+      _extract_instances((a, (b, c)) -> [(a[0], (b[0], c[0])), (a[1], (b[1], c[1])), ...]
+    """
+    if isinstance(x, (np.ndarray, mx.nd.NDArray)):
+        for i in range(x.shape[0]):
+            # yield x[i: i + 1]
+            yield x[i]
+    elif isinstance(x, tuple):
+        for m in zip(*[_extract_instances(y) for y in x]):
+            yield tuple([r for r in m])
+    elif x is None:
+        while True:
+            yield None
+    else:
+        assert False
+
+
+class ForecastGenerator:
+    """
+    Classes used to bring the output of a network into a class.
+    """
+
+    def __call__(
+        self,
+        inference_data_loader: InferenceDataLoader,
+        prediction_net: BlockType,
+        input_names: List[str],
+        freq: str,
+        output_transform: Optional[OutputTransform],
+        num_eval_samples: Optional[int],
+        **kwargs
+    ) -> Iterator[Forecast]:
+        raise NotImplementedError()
+
+
+class DistributionForecastGenerator(ForecastGenerator):
+    @validated()
+    def __init__(self, distr_output: DistributionOutput) -> None:
+        self.distr_output = distr_output
+
+    def __call__(
+        self,
+        inference_data_loader: InferenceDataLoader,
+        prediction_net: BlockType,
+        input_names: List[str],
+        freq: str,
+        output_transform: Optional[OutputTransform],
+        num_eval_samples: Optional[int],
+        **kwargs
+    ) -> Iterator[DistributionForecast]:
+        for batch in inference_data_loader:
+            inputs = [batch[k] for k in input_names]
+            outputs = prediction_net(*inputs)
+            if output_transform is not None:
+                outputs = output_transform(batch, outputs)
+            if num_eval_samples:
+                log_once(
+                    "Forecast is not sample based. Ignoring parameter `num_eval_samples` from predict method."
```

Comment:
We've been calling this `num_eval_samples` everywhere, but I think this is confusing: if anything, I would call this `num_samples` in this context (and in many other places, but outside the scope of this PR). Also, the `"predict"` method is out of context.

My opinion is that this should be hidden in the `**kwargs`, and only be explicitly exposed in `SampleForecastGenerator`, but it's not a big deal.